include/boost/math/statistics/detail/single_pass.hpp

```diff
@@ -0,0 +1,384 @@
+            i += 1;
+        }
+
+        const ReturnType num1 = ReturnType(elements - (elements %4))/ReturnType(4);
+        const ReturnType num2 = num1 + ReturnType(elements % 4);
+
+        for (auto it = end; it != last; ++it)
+        {
+            mu[3] += (*it-mu[3])/i;
+            i += 1;
+        }
+
+        return (num1 * std::valarray<ReturnType>(mu[std::slice(0,3,1)]).sum() + num2 * mu[3]) / ReturnType(elements);
+    }
+    else
+    {
+        auto it = first;
+        ReturnType mu = *it;
+        ReturnType i = 2;
+        while(++it != last)
+        {
+            mu += (*it - mu)/i;
+            i += 1;
+        }
+
+        return mu;
+    }
+}
+
+// Higham, Accuracy and Stability, equation 1.6a and 1.6b:
+// Calculates Mean, M2, and variance
+template<typename ReturnType, typename ForwardIterator>
+ReturnType variance_sequential_impl(ForwardIterator first, ForwardIterator last)
+{
+    using Real = std::tuple_element_t<0, ReturnType>;
+
+    Real M = *first;
+    Real Q = 0;
+    Real k = 2;
+    Real M2 = 0;
+
+    for(auto it = std::next(first); it != last; ++it)
+    {
+        Real tmp = (*it - M) / k;
+        Real delta_1 = *it - M;
+        Q += k*(k-1)*tmp*tmp;
+        M += tmp;
+        k += 1;
+        Real delta_2 = *it - M;
+        M2 += delta_1 * delta_2;
+    }
+
+    return std::make_tuple(M, M2, Q/(k-1));
+}
+
+// Global thread counter required for recursive calls. Reset before each parallel method is called
+static std::atomic<unsigned> thread_counter {1};
+
+// http://i.stanford.edu/pub/cstr/reports/cs/tr/79/773/CS-TR-79-773.pdf
+// http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.214.8508&rep=rep1&type=pdf
+template<typename ReturnType, typename ForwardIterator>
+ReturnType parallel_variance_impl(ForwardIterator first, ForwardIterator last)
+{
+    using Real = typename std::iterator_traits<ForwardIterator>::value_type;
+    static unsigned num_threads {std::thread::hardware_concurrency()};
+
+    const auto elements {std::distance(first, last)};
+    const auto range_a {std::floor(elements / 2)};
+    const auto range_b {elements - range_a};
+
+    thread_counter.fetch_add(2);
+    auto future_a {std::async(std::launch::async, [first, range_a]() -> ReturnType
+    {
+        if(thread_counter + 2 <= num_threads && range_a > 10)
+        {
+            return parallel_variance_impl<ReturnType>(first, std::next(first, range_a));
+        }
+        else
+        {
+            return variance_sequential_impl<ReturnType>(first, std::next(first, range_a));
+        }
+    })};
+    auto future_b {std::async(std::launch::async, [first, last, range_a]() -> ReturnType
+    {
+        if(thread_counter + 2 <= num_threads && range_a > 10)
+        {
+            return parallel_variance_impl<ReturnType>(std::next(first, range_a), last);
+        }
+        else
+        {
+            return variance_sequential_impl<ReturnType>(std::next(first, range_a), last);
+        }
+    })};
```

Comment:
Are you sure this needs to be done via recursion? It looks like the only thing that buys you is the ability to use the formula to pool two variances to pool all the variances. But the Pebay paper shows how to pool n variances without issue.

Reply:
It is done via recursion because I found success using Chan's pairwise algorithms. I did not try Pebay's method.

Reply:
I have no problem with Chan's pairwise algorithm, but if it requires a global it just isn't going to work.